arngll/src/frame_info.rs

```diff
@@ -0,0 +1,690 @@
+
+        if let Some(kid) = self.kid {
+            write!(f, " KID=0x{:02X}",kid)?;
+        }
+
+        write!(f, " MIC={:?}",self.mic)?;
+        write!(f, "}}")
+    }
+}
+
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct FrameInfo {
+    pub frame_type: FrameType,
+    pub ack_requested: bool,
+    pub is_from_relay: bool,
+    pub network_id: Option<NetworkId>,
+    pub dst_addr: HamAddr,
+    pub src_addr: HamAddr,
+    pub rly_addr: Option<HamAddr>,
+    pub sec_info: Option<SecInfo>,
+    pub ack_crc: u16,
+}
+
+impl Debug for FrameInfo {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f,"{{{:?}", self.frame_type)?;
+
+        if self.ack_requested {
+            write!(f," AckReq")?;
+        }
+
+        if self.is_from_relay {
+            write!(f," FromRly")?;
+        }
+
+        if let Some(netid) = self.network_id {
+            write!(f," NetId={:?}", netid)?;
+        }
+
+        if !self.dst_addr.is_empty() {
+            write!(f," Dst={}", self.dst_addr)?;
+        }
+
+        write!(f," Src={}", self.src_addr)?;
+
+        if self.frame_type == FrameType::Ack {
+            write!(f," AckCrc=0x{:04X}", self.ack_crc)?;
+        }
+
+        if let Some(rly_addr) = self.rly_addr {
+            write!(f," Rly={}", rly_addr)?;
+        }
+
+        if let Some(secinfo) = &self.sec_info {
+            write!(f," SecInfo={:?}", secinfo)?;
+        }
+
+        write!(f,"}}")
+    }
+}
+
+impl FrameInfo {
+    pub const EMPTY: FrameInfo = FrameInfo {
+        frame_type: FrameType::Data,
+        ack_requested: false,
+        is_from_relay: false,
+        network_id: None,
+        dst_addr: HamAddr::EMPTY,
+        src_addr: HamAddr::EMPTY,
+        rly_addr: None,
+        sec_info: None,
+        ack_crc: 0,
+    };
+
+    pub fn ack_calc(&self, payload: &[u8]) -> Option<(u16,HamAddr)> {
+        if self.ack_requested {
+            let ack_sender = if let (Some(rly_addr),true) = (self.rly_addr, self.is_from_relay) {
+                rly_addr
+            } else {
+                self.dst_addr
+            };
+            Some((self
+                    .bytes_with_payload(payload)
+                    .fold(X25.digest(), |mut digest, x| {
+                        digest.update(&[x]);
+                        digest
+                    })
+                    .finalize(),
+                  ack_sender,
+            ))
+        } else {
+            None
+        }
+    }
+
+    pub fn generate_ack_frame(&self, payload: &[u8]) -> Option<FrameInfo> {
+        if let Some((ack_crc, src_addr)) = self.ack_calc(payload) {
+            Some(FrameInfo {
+                frame_type: FrameType::Ack,
+                src_addr,
+                ack_crc,
+                .. FrameInfo::EMPTY
+            })
+        } else {
+            None
+        }
+    }
+    
+    pub fn try_from_bytes(frame: &[u8]) -> Result<(FrameInfo, &[u8]), Error> {
```

Comment:
Looking at some of this, I wonder if it may make sense to write it as a bitstruct (https://github.com/dancrossnyc/rust-bitstruct. To give credit where due, while I host it, this is really from Glenn Griffin at el Goog, and we used it on My Last Project there). Then we read into a u32 and pick the fields out from there, or define methods on the resulting struct to do things like, say, return `dst_len`. Probably something to consider in a follow-up, if at all.